Sprint 7 solution in file manager

zlov82 opened the pull request.

zlov82 committed:
fix: Запись и чтение в файл данных о задачах (через FileBackedTaskManager) + рефактор InMemoryTaskManager для корректной загрузки данных из файла через штатные методы добавления задач

 -

zlov82 committed:
fix: В FileBackedTaskManager замена построчного чтения файла на разовое считываение

 -

zlov82 committed:
fix: Конструктор в ManagerSaveException + изменение вызова данного исключения

 -

zlov82 committed:
feature: Добавлены юнит-тесты на чтение-запись в файл FileBackedTaskManagerTest

 -

zlov82 committed:
fix: Удалил отладочные данные из main

 -

zlov82 committed:
fix: Reformat Code

YazykovMikhail commented on src/ru/yandex/javacourse/kanban/manager/FileBackedTaskManager.java:
```diff
@@ -0,0 +1,196 @@
+        return fileBackedTaskManager;
+    }
+
+    private final String pathToFileDb;
```
Поля обьявляются первыми в классе
Храни сам обект File
private final File file;

YazykovMikhail commented on src/ru/yandex/javacourse/kanban/manager/FileBackedTaskManager.java:
```diff
@@ -0,0 +1,196 @@
+
+    @Override
+    public int addNewTask(Task task) {
+        int ret = super.addNewTask(task);
```
не понятно что такое ret
id - понятно сразу)

YazykovMikhail commented on src/ru/yandex/javacourse/kanban/manager/FileBackedTaskManager.java:
```diff
@@ -0,0 +1,196 @@
+        }
+    }
+
+    private static void fromString(String fileLine, FileBackedTaskManager manager) {
```
Метод назыввется fromString - он должен что то возвращат ьиз строки, в данном случае задачу: 

```
public static Task taskFromString(String value) {
		final String[] values = value.split(",");
		final int id = Integer.parseInt(values[0]);
		final TaskType type = TaskType.valueOf(values[1]);
		final String name = values[2];
		final TaskStatus status = TaskStatus.valueOf(values[3]);
		final String description = values[4];
		if (type == TaskType.TASK) {
			return new Task(id, name, description, status);
		}
		if (type == TaskType.SUBTASK) {
			final int epicId = Integer.parseInt(values[5]);
			return new Subtask(id, name, description, status, epicId);
		}

		return new Epic(id, name, description, status);
	}
```

А добавлять их в хранилище - это дело метода который вызывает текущей, каждый должен отвечать за свое

YazykovMikhail commented on src/ru/yandex/javacourse/kanban/tasks/Task.java:
```diff
@@ -70,15 +69,23 @@ public void setStatus(TaskStatus status) {
 
```
Добавь поле TaskType taskType
или еще проще сделай метод для каждой задачи 
public TaskType getType() {
		return TaskType.TASK;
	}
который будет возвращать свой тип.

YazykovMikhail commented on src/ru/yandex/javacourse/kanban/tasks/Task.java:
```diff
@@ -70,15 +69,23 @@ public void setStatus(TaskStatus status) {
 
     @Override
     public String toString() {
```
Не стоит использовать базовый метод toString() в бизнес процессах, этот метод много где используется самой Джавой, для бизнес процесса стоит сделать отдельный метод который будет делать по конкретному шаблону, который отвечает бизнес треованиям

YazykovMikhail commented on src/ru/yandex/javacourse/kanban/manager/FileBackedTaskManager.java:
```diff
@@ -0,0 +1,196 @@
+            throw new ManagerSaveException(e);
+        }
+    }
+
```
Стоит сделать отдельный метод toString который используется для бизнес процесов, например вот простой:
```
	public static String toString(Task task) {
		return task.getId() + "," + task.getType() + "," + task.getName() + "," + task.getStatus() + "," + task.getDescription() + "," + (task.getType().equals(TaskType.SUBTASK) ? task.getEpicId() : "");
	}
```

zlov82 replied:
task.getEpicId() - метод не находиться в таком виде, несмотря на то, что он публичный

YazykovMikhail commented on src/ru/yandex/javacourse/kanban/manager/FileBackedTaskManager.java:
```diff
@@ -0,0 +1,196 @@
+        return super.getHistory();
+    }
+
+    private void save() {
```
Есть пару моментов которые можно улучшить:
1) Использовать буферизацию - экономит ресурсы
2) Не стоит использовать "\n" - в разных операционных системах перенос строки разный. Используй системный перенос строки

СПОЙЛЕР:
```
protected void save() {
		try (BufferedWriter writer = new BufferedWriter(new FileWriter(file))) {
			writer.write(CSVTaskFormat.getHeader());
			writer.newLine();

			for (Map.Entry<Integer, Task> entry : tasks.entrySet()) {
				final Task task = entry.getValue();
				writer.write(CSVTaskFormat.toString(task));
				writer.newLine();
			}

			for (Map.Entry<Integer, Subtask> entry : subtasks.entrySet()) {
				final Task task = entry.getValue();
				writer.write(CSVTaskFormat.toString(task));
				writer.newLine();
			}

			for (Map.Entry<Integer, Epic> entry : epics.entrySet()) {
				final Task task = entry.getValue();
				writer.write(CSVTaskFormat.toString(task));
				writer.newLine();
			}

			writer.newLine();
		} catch (IOException e) {
			throw new ManagerSaveException("Can't save to file: " + file.getName(), e);
		}
	}
```

YazykovMikhail commented on src/ru/yandex/javacourse/kanban/manager/FileBackedTaskManager.java:
```diff
@@ -0,0 +1,196 @@
+
+public class FileBackedTaskManager extends InMemoryTaskManager {
+
+    public static FileBackedTaskManager loadFromFile(File file) throws ManagerSaveException {
```
В целом ок, но есть что улучшить
1) можно не описывать в сигнатуре метода throws ManagerSaveException - это исключение от RuntimeException если оно произошло то по нашей логике что то сломало, специально говорить о том что его нужно обработать не нужно
2) Не востанавливается счетчик идентификатора
3) Можно не переделывать старые методы add что бы в зависимоси от наличия id они уменли добавлять - это может создать баги, когда с фронта будут приходить новые задачи уже с дентификатором. Добавляй прям сразу их в хранилище напрямую
4) Метод fromString должен создавать задачи из текста, а он к тому же еще и добавляет, каждый мтод должен отвечать за свою часть работы

СПОЙЛЕР:

```
public static FileBackedTaskManager loadFromFile(File file) {
		final FileBackedTaskManager taskManager = new FileBackedTaskManager(file);
		try {
			final String csv = Files.readString(file.toPath());
			final String[] lines = csv.split(System.lineSeparator());
			int generatorId = 0;
			for (int i = 1; i < lines.length; i++) {
				String line = lines[i];
				if (line.isEmpty()) {
					break;
				}
				final Task task = CSVTaskFormat.taskFromString(line);
				final int id = task.getId();
				if (id > generatorId) {
					generatorId = id;
				}
				taskManager.addAnyTask(task);
			}
			for (Map.Entry<Integer, Subtask> e : taskManager.subtasks.entrySet()) {
				final Subtask subtask = e.getValue();
				final Epic epic = taskManager.epics.get(subtask.getEpicId());
				epic.addSubtaskId(subtask.getId());
			}
			taskManager.uniqueId = generatorId;
		} catch (IOException e) {
			throw new ManagerSaveException("Can't read form file: " + file.getName(), e);
		}
		return taskManager;
	}

```

доп метод:
```
protected void addAnyTask(Task task) {
		final int id = task.getId();
		switch (task.getType()) {
			case TASK:
				tasks.put(id, task);
				break;
			case SUBTASK:
				subtasks.put(id, (Subtask) task);
				break;
			case EPIC:
				epics.put(id, (Epic) task);
				break;
		}
	}
```

zlov82 replied:
Счетчик идентификатора у меня восстанавливался в private static void fromString
       int taskId = Integer.parseInt(lines[0]);
        if (manager.uniqueId < taskId) {
            manager.uniqueId = taskId;
        }

YazykovMikhail commented on src/ru/yandex/javacourse/kanban/manager/FileBackedTaskManager.java:
```diff
@@ -0,0 +1,196 @@
+        save();
+    }
+
+    @Override
```
Все методы переопреелять не нужно, только те которые влияют на состояние задач. 
Раз тут состояние не меняется, то и сохранять не нужено и соответсвенно переопределять, ведь он в итоге делает тоже самое
Стоит убрать все методы не меняющие состояние

YazykovMikhail commented on src/ru/yandex/javacourse/kanban/manager/Managers.java:
```diff
@@ -1,9 +1,12 @@
     public static TaskManager getDefault() {
-        return new InMemoryTaskManager();
+        //return new InMemoryTaskManager();
+        return new FileBackedTaskManager(new File("res.csv"));
```
Положи файл в папку
"resources/task.csv"

zlov82 committed:
fix: файл БД перемешён в resources/task.csv

zlov82 committed:
fix:
 - убрал лишние переопределения в FileBackedTaskManager
 - срока записи для тасков генериться из отдельного метода класса FileBackedTaskManager
 - сохранение с файл через буффер

zlov82 committed:
fix: скорректирован метод загрузки loadFromFile

zlov82 committed:
add: каталог ресурсов

YazykovMikhail commented on src/ru/yandex/javacourse/kanban/manager/FileBackedTaskManager.java:
```diff
@@ -0,0 +1,199 @@
+                }
+            }
+        } catch (IOException e) {
+            throw new ManagerSaveException(e);
```
стоит добавлять какое то описание к исключение, например как было в предыдущем ревью

YazykovMikhail replied:
На будущее

YazykovMikhail commented on src/ru/yandex/javacourse/kanban/manager/FileBackedTaskManager.java:
```diff
@@ -0,0 +1,199 @@
+                writer.newLine();
+            }
+        } catch (IOException e) {
+            throw new ManagerSaveException(e);
```
Стоит добавлять текст в исключениях, с опсианием/причиной
